packages/json-schema/test/integration/build-schema.test.ts

```diff
@@ -0,0 +1,216 @@
+      class Empty {}
+      class NoModelMeta {
+        @property() foo: string;
+        bar: number;
+      }
+      expect(modelToJsonDef(Empty)).to.eql({});
+      expect(modelToJsonDef(NoModelMeta)).to.eql({});
+    });
+    it('properly converts string, number, and boolean properties', () => {
+      @model()
+      class TestModel {
+        @property() str: string;
+        @property() num: number;
+        @property() bool: boolean;
+      }
+
+      const jsonDef = modelToJsonDef(TestModel);
+      expect(jsonDef.properties).to.deepEqual({
+        str: {
+          type: 'string',
+        },
+        num: {
+          type: 'number',
+        },
+        bool: {
+          type: 'boolean',
+        },
+      });
+    });
+    it('properly converts object properties', () => {
+      @model()
+      class TestModel {
+        @property() obj: object;
+      }
+
+      const jsonDef = modelToJsonDef(TestModel);
+      expect(jsonDef.properties).to.deepEqual({
+        obj: {
+          type: 'object',
+        },
+      });
+    });
+    it('properly converts custom type properties', () => {
+      class CustomType {
+        prop: string;
+      }
+
+      @model()
+      class TestModel {
+        @property() cusType: CustomType;
+      }
+
+      const jsonDef = modelToJsonDef(TestModel);
+      expect(jsonDef.properties).to.deepEqual({
+        cusType: {
+          $ref: '#definitions/CustomType',
+        },
+      });
+    });
+    it('properly converts primitive arrays properties', () => {
+      @model()
+      class TestModel {
+        @property.array(Number) numArr: number[];
+      }
+
+      const jsonDef = modelToJsonDef(TestModel);
+      expect(jsonDef.properties).to.deepEqual({
+        numArr: {
+          type: 'array',
+          items: {
+            type: 'number',
+          },
+        },
+      });
+    });
+    it('properly converts custom type arrays properties', () => {
+      class CustomType {
+        prop: string;
+      }
+
+      @model()
+      class TestModel {
+        @property.array(CustomType) cusArr: CustomType[];
+      }
+
+      const jsonDef = modelToJsonDef(TestModel);
+      expect(jsonDef.properties).to.deepEqual({
+        cusArr: {
+          type: 'array',
+          items: {
+            $ref: '#definitions/CustomType',
+          },
+        },
+      });
+    });
+    describe('toJSONProperty', () => {
+      class Bar {
+        barA: number;
+      }
+      @model()
+      class Foo {
+        @property() str: string;
+        @property() num: number;
+        @property() bool: boolean;
+        @property() obj: object;
+        @property() nul: null;
+        @property() undef: undefined;
+        @property.array(String) arrStr: string[];
+        @property() bar: Bar;
+        @property.array(Bar) arrBar: Bar[];
+      }
+      // tslint:disable-next-line:no-any
+      let meta: any;
+      // tslint:disable-next-line:no-any
+      let propMeta: any;
```

Comment:
Similarly here. I think we may want to use [`PropertyMap `](https://github.com/strongloop/loopback-next/blob/b19635d821c6beb6dd1c9597e3153abc073da1e4/packages/repository/src/decorators/model.ts#L22) that's private right now?